## **Detecting and Monitoring Land Use Transformation in Córdoba**

### Import necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

In [2]:
import ee

# Initialize GEE
ee.Authenticate()
# Initialize GEE with your GCP project ID
ee.Initialize(project='gee-land-use-cordoba')

In [3]:
# Load a sample dataset
dem = ee.Image('USGS/SRTMGL1_003')

# Print metadata
print(dem.getInfo())

{'type': 'Image', 'bands': [{'id': 'elevation', 'data_type': {'type': 'PixelType', 'precision': 'int', 'min': -32768, 'max': 32767}, 'dimensions': [1296001, 417601], 'crs': 'EPSG:4326', 'crs_transform': [0.0002777777777777778, 0, -180.0001388888889, 0, -0.0002777777777777778, 60.00013888888889]}], 'version': 1641990767055141, 'id': 'USGS/SRTMGL1_003', 'properties': {'system:visualization_0_min': '0.0', 'type_name': 'Image', 'keywords': ['dem', 'elevation', 'geophysical', 'nasa', 'srtm', 'topography', 'usgs'], 'thumb': 'https://mw1.google.com/ges/dd/images/SRTM90_V4_thumb.png', 'description': '<p>The Shuttle Radar Topography Mission (SRTM, see <a href="https://onlinelibrary.wiley.com/doi/10.1029/2005RG000183/full">Farr\net al. 2007</a>)\ndigital elevation data is an international research effort that\nobtained digital elevation models on a near-global scale. This\nSRTM V3 product (SRTM Plus) is provided by NASA JPL\nat a resolution of 1 arc-second (approximately 30m).</p><p>This dataset

In [4]:
import geemap

# Define AoIs with provided coordinates
aoi_1 = ee.Geometry.Rectangle([-64.53518510984365, -32.09793491720308, -64.36758169478006, -31.98599080824592])  # Calmayo
aoi_2 = ee.Geometry.Rectangle([-64.19504913677835, -30.67255559541076, -63.92899759153784, -30.43974550819874])  # Las Peñas
aoi_3 = ee.Geometry.Rectangle([-64.8501456212574, -32.06252532553296, -64.70773302600543, -31.93373503424498])  # Villa Alpina

# Create an interactive map
Map = geemap.Map()

# Add AoIs to the map
Map.addLayer(aoi_1, {'color': 'red'}, 'Calmayo')
Map.addLayer(aoi_2, {'color': 'blue'}, 'Las Peñas')
Map.addLayer(aoi_3, {'color': 'green'}, 'Villa Alpina')

# Center the map on Córdoba
Map.centerObject(aoi_1, 10)
Map

Map(center=[-32.041979015931354, -64.45138340231044], controls=(WidgetControl(options=['position', 'transparen…

#### Load and Preprocess Satellite Imagery

In [5]:
def load_sentinel2(start_date, end_date, region):
    return (
        ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED')
        .filterDate(start_date, end_date)
        .filterBounds(region)
        .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 30))  # Max 30% cloud coverage
        .select(['B4', 'B3', 'B2', 'B8', 'B11'])  # Select necessary bands before applying media
    )

#### Load Images for Summer and Winter

In [6]:
def load_landsat8(start_date, end_date, region):
    return (
        ee.ImageCollection('LANDSAT/LC08/C02/T1_L2')
        .filterDate(start_date, end_date)
        .filterBounds(region)
        .filter(ee.Filter.lt('CLOUD_COVER', 30))  # Max 30% cloud coverage
        .select(['SR_B2', 'SR_B3', 'SR_B4', 'SR_B5', 'SR_B6', 'SR_B7'])  # Updated band names for Collection 2, Level 2
    )

In [7]:
# Load Sentinel-2 images for all AoIs and periods (3-year timeframe)
summer_sentinel2_aoi1 = load_sentinel2('2017-12-01', '2020-03-31', aoi_1)
winter_sentinel2_aoi1 = load_sentinel2('2017-06-01', '2019-09-30', aoi_1)

summer_sentinel2_aoi2 = load_sentinel2('2017-12-01', '2020-03-31', aoi_2)
winter_sentinel2_aoi2 = load_sentinel2('2017-06-01', '2019-09-30', aoi_2)

summer_sentinel2_aoi3 = load_sentinel2('2017-12-01', '2020-03-31', aoi_3)
winter_sentinel2_aoi3 = load_sentinel2('2017-06-01', '2019-09-30', aoi_3)

# Load Landsat 8 images for all AoIs and periods (3-year timeframe)
summer_landsat8_aoi1 = load_landsat8('2013-12-01', '2016-03-31', aoi_1)
winter_landsat8_aoi1 = load_landsat8('2013-06-01', '2015-09-30', aoi_1)

summer_landsat8_aoi2 = load_landsat8('2013-12-01', '2016-03-31', aoi_2)
winter_landsat8_aoi2 = load_landsat8('2013-06-01', '2015-09-30', aoi_2)

summer_landsat8_aoi3 = load_landsat8('2013-12-01', '2016-03-31', aoi_3)
winter_landsat8_aoi3 = load_landsat8('2013-06-01', '2015-09-30', aoi_3)

In [8]:
# Store Sentinel-2 collections in a list for easier access
sentinel2_collections = [
    (summer_sentinel2_aoi1, winter_sentinel2_aoi1),
    (summer_sentinel2_aoi2, winter_sentinel2_aoi2),
    (summer_sentinel2_aoi3, winter_sentinel2_aoi3),
]

# Store Landsat 8 collections in a list for easier access
landsat8_collections = [
    (summer_landsat8_aoi1, winter_landsat8_aoi1),
    (summer_landsat8_aoi2, winter_landsat8_aoi2),
    (summer_landsat8_aoi3, winter_landsat8_aoi3),
]

In [9]:
# Example for Sentinel-2 collections
for i, (summer_collection, winter_collection) in enumerate(sentinel2_collections):
    aoi_name = f"aoi_{i + 1}"

    # Check summer collection
    summer_size = summer_collection.size().getInfo()  # Requires ee.Initialize()
    print(f"Sentinel-2 Summer ({aoi_name}): {summer_size} images found.")
    if summer_size > 0:
        summer_collection = summer_collection.median()
    else:
        print(f"Warning: No summer images for {aoi_name}!")

    # Check winter collection
    winter_size = winter_collection.size().getInfo()
    print(f"Sentinel-2 Winter ({aoi_name}): {winter_size} images found.")
    if winter_size > 0:
        winter_collection = winter_collection.median()
    else:
        print(f"Warning: No winter images for {aoi_name}!")

Sentinel-2 Summer (aoi_1): 100 images found.
Sentinel-2 Winter (aoi_1): 60 images found.
Sentinel-2 Summer (aoi_2): 185 images found.
Sentinel-2 Winter (aoi_2): 105 images found.
Sentinel-2 Summer (aoi_3): 100 images found.
Sentinel-2 Winter (aoi_3): 60 images found.


In [10]:
# Check Landsat 8 collections
for i, (summer_collection, winter_collection) in enumerate(landsat8_collections):
    aoi_name = f"aoi_{i + 1}"

    # Process summer collection
    summer_size = summer_collection.size().getInfo()
    print(f"Landsat 8 Summer ({aoi_name}): {summer_size} images found.")
    if summer_size > 0:
        summer_collection = summer_collection.median()
    else:
        print(f"Warning: Summer Landsat 8 collection is empty for {aoi_name}.")

    # Process winter collection
    winter_size = winter_collection.size().getInfo()
    print(f"Landsat 8 Winter ({aoi_name}): {winter_size} images found.")
    if winter_size > 0:
        winter_collection = winter_collection.median()
    else:
        print(f"Warning: Winter Landsat 8 collection is empty for {aoi_name}.")

Landsat 8 Summer (aoi_1): 22 images found.
Landsat 8 Winter (aoi_1): 21 images found.
Landsat 8 Summer (aoi_2): 21 images found.
Landsat 8 Winter (aoi_2): 19 images found.
Landsat 8 Summer (aoi_3): 51 images found.
Landsat 8 Winter (aoi_3): 51 images found.


In [11]:
print(type(summer_sentinel2_aoi1).__name__)  # Should print 'ImageCollection'
print(type(winter_sentinel2_aoi1).__name__)  # Should print 'ImageCollection'

ImageCollection
ImageCollection


In [12]:
# Resample Landsat 8 data to 10 meters to match Sentinel-2 resolution
def resample_landsat(image):
    return image.resample('bilinear').reproject(crs='EPSG:4326', scale=10)

summer_l8_aoi1 = resample_landsat(summer_landsat8_aoi1.median())
winter_l8_aoi1 = resample_landsat(winter_landsat8_aoi1.median())

summer_l8_aoi2 = resample_landsat(summer_landsat8_aoi2.median())
winter_l8_aoi2 = resample_landsat(winter_landsat8_aoi2.median())

summer_l8_aoi3 = resample_landsat(summer_landsat8_aoi3.median())
winter_l8_aoi3 = resample_landsat(winter_landsat8_aoi3.median())

### Data Preprocessing (Data Harmonization)

**Sentinel-2 Bands**
* B2: Blue (10m)

* B3: Green (10m)

* B4: Red (10m)

* B8: Near Infrared (NIR, 10m)

* B11: Shortwave Infrared (SWIR, 20m)

**Landsat 8 Bands**
* SR_B2: Blue (30m)

* SR_B3: Green (30m)

* SR_B4: Red (30m)

* SR_B5: Near Infrared (NIR, 30m)

* SR_B6: Shortwave Infrared 1 (SWIR1, 30m)

* SR_B7: Shortwave Infrared 2 (SWIR2, 30m)

Combining Sentinel-2 and Landsat 8 data into a single dataset for each AOI


**Including Additional Indices**

**Common Indices**:

NDWI (Normalized Difference Water Index): For water body detection.

* Sentinel-2: (B3 - B8) / (B3 + B8)

* Landsat 8: (SR_B3 - SR_B5) / (SR_B3 + SR_B5)

NDBI (Normalized Difference Built-Up Index): For urban area detection.

* Sentinel-2: (B11 - B8) / (B11 + B8)

* Landsat 8: (SR_B6 - SR_B5) / (SR_B6 + SR_B5)

EVI (Enhanced Vegetation Index): For vegetation monitoring.

* Sentinel-2: 2.5 * (B8 - B4) / (B8 + 6 * B4 - 7.5 * B2 + 1)

* Landsat 8: 2.5 * (SR_B5 - SR_B4) / (SR_B5 + 6 * SR_B4 - 7.5 * SR_B2 + 1)



##### Processing Sentinel-2 Data for All AOIs

Computing medians and add spectral indices (NDVI, NDWI, NDBI, EVI) for Sentinel-2 data

In [13]:
def process_s2(summer_collection, winter_collection):
    # Compute medians
    summer_median = summer_collection.median() if summer_collection.size().getInfo() > 0 else None
    winter_median = winter_collection.median() if winter_collection.size().getInfo() > 0 else None

    # Add indices for summer median
    if summer_median:
        summer_median = summer_median.addBands([
            summer_median.normalizedDifference(['B8', 'B4']).rename('NDVI'),  # NDVI
            summer_median.normalizedDifference(['B3', 'B8']).rename('NDWI'),  # NDWI
            summer_median.normalizedDifference(['B11', 'B8']).rename('NDBI'),  # NDBI
            summer_median.expression(
                '2.5 * (NIR - RED) / (NIR + 6 * RED - 7.5 * BLUE + 1)', {
                    'NIR': summer_median.select('B8'),
                    'RED': summer_median.select('B4'),
                    'BLUE': summer_median.select('B2')
                }
            ).rename('EVI')  # EVI
        ])

    # Add indices for winter median
    if winter_median:
        winter_median = winter_median.addBands([
            winter_median.normalizedDifference(['B8', 'B4']).rename('NDVI'),  # NDVI
            winter_median.normalizedDifference(['B3', 'B8']).rename('NDWI'),  # NDWI
            winter_median.normalizedDifference(['B11', 'B8']).rename('NDBI'),  # NDBI
            winter_median.expression(
                '2.5 * (NIR - RED) / (NIR + 6 * RED - 7.5 * BLUE + 1)', {
                    'NIR': winter_median.select('B8'),
                    'RED': winter_median.select('B4'),
                    'BLUE': winter_median.select('B2')
                }
            ).rename('EVI')  # EVI
        ])

    return summer_median, winter_median

##### Processing Landsat 8 Data for All AOIs

Compute medians and add spectral indices (NDVI, NDWI, NDBI, EVI) for Landsat 8 data.

In [14]:
def process_l8(summer_collection, winter_collection):
    # Compute medians
    summer_median = summer_collection.median() if summer_collection.size().getInfo() > 0 else None
    winter_median = winter_collection.median() if winter_collection.size().getInfo() > 0 else None

    # Add indices for summer median
    if summer_median:
        summer_median = summer_median.addBands([
            summer_median.normalizedDifference(['SR_B5', 'SR_B4']).rename('NDVI'),  # NDVI
            summer_median.normalizedDifference(['SR_B3', 'SR_B5']).rename('NDWI'),  # NDWI
            summer_median.normalizedDifference(['SR_B6', 'SR_B5']).rename('NDBI'),  # NDBI
            summer_median.expression(
                '2.5 * (NIR - RED) / (NIR + 6 * RED - 7.5 * BLUE + 1)', {
                    'NIR': summer_median.select('SR_B5'),
                    'RED': summer_median.select('SR_B4'),
                    'BLUE': summer_median.select('SR_B2')
                }
            ).rename('EVI')  # EVI
        ])

    # Add indices for winter median
    if winter_median:
        winter_median = winter_median.addBands([
            winter_median.normalizedDifference(['SR_B5', 'SR_B4']).rename('NDVI'),  # NDVI
            winter_median.normalizedDifference(['SR_B3', 'SR_B5']).rename('NDWI'),  # NDWI
            winter_median.normalizedDifference(['SR_B6', 'SR_B5']).rename('NDBI'),  # NDBI
            winter_median.expression(
                '2.5 * (NIR - RED) / (NIR + 6 * RED - 7.5 * BLUE + 1)', {
                    'NIR': winter_median.select('SR_B5'),
                    'RED': winter_median.select('SR_B4'),
                    'BLUE': winter_median.select('SR_B2')
                }
            ).rename('EVI')  # EVI
        ])

    return summer_median, winter_median

In [15]:
def fuse_data(s2_img, l8_img):
    # Combine all bands and indices
    return ee.Image.cat([
        s2_img.select(['B2', 'B3', 'B4', 'B8', 'B11']),  # Sentinel-2 bands
        s2_img.select(['NDVI', 'NDWI', 'NDBI', 'EVI']),  # Sentinel-2 indices
        l8_img.select(['SR_B2', 'SR_B3', 'SR_B4', 'SR_B5', 'SR_B6', 'SR_B7']),  # Landsat 8 bands
        l8_img.select(['NDVI', 'NDWI', 'NDBI', 'EVI'])  # Landsat 8 indices
    ]).rename([
        'S2_BLUE', 'S2_GREEN', 'S2_RED', 'S2_NIR', 'S2_SWIR',  # Sentinel-2 bands
        'S2_NDVI', 'S2_NDWI', 'S2_NDBI', 'S2_EVI',  # Sentinel-2 indices
        'L8_BLUE', 'L8_GREEN', 'L8_RED', 'L8_NIR', 'L8_SWIR1', 'L8_SWIR2',  # Landsat 8 bands
        'L8_NDVI', 'L8_NDWI', 'L8_NDBI', 'L8_EVI'  # Landsat 8 indices
    ])

In [16]:
# Process and update the variables with the median images
summer_s2_aoi1, winter_s2_aoi1 = process_s2(summer_sentinel2_aoi1, winter_sentinel2_aoi1)
summer_s2_aoi2, winter_s2_aoi2 = process_s2(summer_sentinel2_aoi2, winter_sentinel2_aoi2)
summer_s2_aoi3, winter_s2_aoi3 = process_s2(summer_sentinel2_aoi3, winter_sentinel2_aoi3)

summer_l8_aoi1, winter_l8_aoi1 = process_l8(summer_landsat8_aoi1, winter_landsat8_aoi1)
summer_l8_aoi2, winter_l8_aoi2 = process_l8(summer_landsat8_aoi2, winter_landsat8_aoi2)
summer_l8_aoi3, winter_l8_aoi3 = process_l8(summer_landsat8_aoi3, winter_landsat8_aoi3)

Fuse Data for All AOIs

In [17]:
# Fuse data for AOI 1
combined_aoi1 = fuse_data(summer_s2_aoi1, summer_l8_aoi1)
# Fuse data for AOI 2
combined_aoi2 = fuse_data(summer_s2_aoi2, summer_l8_aoi2)
# Fuse data for AOI 3
combined_aoi3 = fuse_data(summer_s2_aoi3, summer_l8_aoi3)

In [18]:
# Verify the Fused Data
print(combined_aoi1.bandNames().getInfo())
print(combined_aoi2.bandNames().getInfo())
print(combined_aoi3.bandNames().getInfo())

['S2_BLUE', 'S2_GREEN', 'S2_RED', 'S2_NIR', 'S2_SWIR', 'S2_NDVI', 'S2_NDWI', 'S2_NDBI', 'S2_EVI', 'L8_BLUE', 'L8_GREEN', 'L8_RED', 'L8_NIR', 'L8_SWIR1', 'L8_SWIR2', 'L8_NDVI', 'L8_NDWI', 'L8_NDBI', 'L8_EVI']
['S2_BLUE', 'S2_GREEN', 'S2_RED', 'S2_NIR', 'S2_SWIR', 'S2_NDVI', 'S2_NDWI', 'S2_NDBI', 'S2_EVI', 'L8_BLUE', 'L8_GREEN', 'L8_RED', 'L8_NIR', 'L8_SWIR1', 'L8_SWIR2', 'L8_NDVI', 'L8_NDWI', 'L8_NDBI', 'L8_EVI']
['S2_BLUE', 'S2_GREEN', 'S2_RED', 'S2_NIR', 'S2_SWIR', 'S2_NDVI', 'S2_NDWI', 'S2_NDBI', 'S2_EVI', 'L8_BLUE', 'L8_GREEN', 'L8_RED', 'L8_NIR', 'L8_SWIR1', 'L8_SWIR2', 'L8_NDVI', 'L8_NDWI', 'L8_NDBI', 'L8_EVI']


Using ESA WorldCover for ground truth labels

In [19]:
# Load ESA WorldCover dataset
worldcover = ee.Image("ESA/WorldCover/v100/2020")

# Clip to AOIs
labels_aoi1 = worldcover.clip(aoi_1)
labels_aoi2 = worldcover.clip(aoi_2)
labels_aoi3 = worldcover.clip(aoi_3)

In [22]:
# Convert ESA WorldCover image to FeatureCollection using stratified sampling
labels_aoi1_fc = labels_aoi1.stratifiedSample(
    numPoints=1000,  # Number of points to sample
    classBand='Map',  # Band containing class labels
    region=aoi_1,     # Region of interest
    scale=10,         # Resolution (10m for Sentinel-2)
    seed=42,          # Random seed for reproducibility
    geometries=True   # Include geometries in the output
)

labels_aoi2_fc = labels_aoi2.stratifiedSample(
    numPoints=1000,
    classBand='Map',
    region=aoi_2,
    scale=10,
    seed=42,
    geometries=True
)

labels_aoi3_fc = labels_aoi3.stratifiedSample(
    numPoints=1000,
    classBand='Map',
    region=aoi_3,
    scale=10,
    seed=42,
    geometries=True
)

Prepare Training Data

In [23]:
# Sample training data from each AOI
samples_aoi1 = combined_aoi1.sampleRegions(
    collection=labels_aoi1_fc,  # Use the FeatureCollection
    properties=['Map'],  # ESA WorldCover uses 'Map' for class labels
    scale=10  # Consistent resolution for both Sentinel-2 and Landsat
)

samples_aoi2 = combined_aoi2.sampleRegions(
    collection=labels_aoi2_fc,
    properties=['Map'],
    scale=10
)

samples_aoi3 = combined_aoi3.sampleRegions(
    collection=labels_aoi3_fc,
    properties=['Map'],
    scale=10
)

# Merge all samples
merged_samples = ee.FeatureCollection([samples_aoi1, samples_aoi2, samples_aoi3]).flatten()

#### Model Training

In [24]:
# Split data into training/validation
training_data = merged_samples.randomColumn('split')
train = training_data.filter(ee.Filter.lt('split', 0.7))  # 70% training
val = training_data.filter(ee.Filter.gte('split', 0.7))  # 30% validation

# Train a classifier (e.g., Random Forest)
classifier = ee.Classifier.smileRandomForest(100).train(
    features=train,
    classProperty='Map',  # ESA WorldCover uses 'Map' for class labels
    inputProperties=combined_aoi1.bandNames()  # Use all bands/indices
)


In [25]:
# Validate accuracy
val_predictions = val.classify(classifier)  # Classify validation data
confusion_matrix = ee.ConfusionMatrix(val_predictions.errorMatrix('Map', 'classification'))
val_accuracy = confusion_matrix.accuracy()
print('Validation Accuracy:', val_accuracy.getInfo())

Validation Accuracy: 0.7425939306358381


In [27]:
# Validate accuracy
val_predictions = val.classify(classifier)  # Classify validation data
confusion_matrix = ee.ConfusionMatrix(val_predictions.errorMatrix('Map', 'classification'))

# Get the confusion matrix as an array
confusion_matrix_array = confusion_matrix.array()

# Extract true positives (TP), false positives (FP), and false negatives (FN) for each class
# Note: GEE's confusion matrix is organized as [actual][predicted]
matrix = confusion_matrix_array.getInfo()  # Convert to a Python list

# Calculate precision, recall, and F1-score for each class
num_classes = len(matrix)
f1_scores = []

for i in range(num_classes):
    # True positives (TP)
    tp = matrix[i][i]

    # False positives (FP)
    fp = sum(matrix[j][i] for j in range(num_classes) if j != i)

    # False negatives (FN)
    fn = sum(matrix[i][j] for j in range(num_classes) if j != i)

    # Precision
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0

    # Recall
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0

    # F1-score
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0

    f1_scores.append(f1)
    print(f"Class {i}: Precision = {precision:.4f}, Recall = {recall:.4f}, F1-score = {f1:.4f}")

# Calculate the weighted average F1-score
class_counts = [sum(matrix[i]) for i in range(num_classes)]  # Number of samples per class
total_samples = sum(class_counts)
weighted_f1 = sum(f1_scores[i] * class_counts[i] for i in range(num_classes)) / total_samples

print(f"Weighted Average F1-score: {weighted_f1:.4f}")

Class 0: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 1: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 2: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 3: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 4: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 5: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 6: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 7: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 8: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 9: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 10: Precision = 0.8112, Recall = 0.8326, F1-score = 0.8217
Class 11: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 12: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 13: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 14: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 15: Precision = 0.0000, Recal

In [28]:
# Combine the labels for all AOIs
labels_combined = ee.ImageCollection([labels_aoi1, labels_aoi2, labels_aoi3]).mosaic()

# Compute the histogram of class labels
histogram = labels_combined.reduceRegion(
    reducer=ee.Reducer.frequencyHistogram(),
    geometry=aoi_1.union(aoi_2).union(aoi_3),  # Combine all AOIs
    scale=10,
    maxPixels=1e13
)

# Print the class distribution
print(histogram.getInfo())

{'Map': {'10': 2188043.07843137, '20': 2289987.807843136, '30': 3952788.1450980776, '40': 3671142.1372548966, '50': 51063.68627450975, '60': 119825.6784313725, '80': 651, '90': 4}}


Frequency of each ESA WorldCover class in the AOIs (Calmayo, Las Peñas, and Villa Alpina)

* "10": 2188043.07843137,  # Trees
* "20": 2289987.807843136, # Shrubland
* "30": 3952788.1450980776,# Grassland
* "40": 3671142.1372548966,# Cropland
* "50": 51063.68627450975, # Built-up
* "60": 119825.6784313725, # Bare / sparse vegetation
* "80": 651,               # Permanent Water Bodies
* "90": 4                  # Herbaceous Wetland

In [30]:
# List of relevant classes
relevant_classes = [10, 20, 30, 40, 50, 60]  # Trees, Shrubland, Grassland, Cropland, Built-up, Bare / Sparse Vegetation

# Filter the training data to include only relevant classes
train_filtered = train.filter(ee.Filter.inList('Map', relevant_classes))

In [33]:
# Train the Random Forest model without class weights
classifier = ee.Classifier.smileRandomForest(
    numberOfTrees=100  # Number of trees
).train(
    features=train_filtered,  # Use balanced training data
    classProperty='Map',       # Class label property
    inputProperties=combined_aoi1.bandNames()  # Input features
)

In [34]:
# Classify the validation data
val_predictions = val.classify(classifier)

# Generate the confusion matrix
confusion_matrix = val_predictions.errorMatrix('Map', 'classification')

# Calculate overall accuracy
val_accuracy = confusion_matrix.accuracy()
print('Validation Accuracy:', val_accuracy.getInfo())

# Extract the confusion matrix as an array
matrix = confusion_matrix.array().getInfo()

# Calculate precision, recall, and F1-score for each class
num_classes = len(matrix)
f1_scores = []

for i in range(num_classes):
    # True positives (TP)
    tp = matrix[i][i]

    # False positives (FP)
    fp = sum(matrix[j][i] for j in range(num_classes) if j != i)

    # False negatives (FN)
    fn = sum(matrix[i][j] for j in range(num_classes) if j != i)

    # Precision
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0

    # Recall
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0

    # F1-score
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0

    f1_scores.append(f1)
    print(f"Class {i}: Precision = {precision:.4f}, Recall = {recall:.4f}, F1-score = {f1:.4f}")

# Calculate the weighted average F1-score
class_counts = [sum(matrix[i]) for i in range(num_classes)]  # Number of samples per class
total_samples = sum(class_counts)
weighted_f1 = sum(f1_scores[i] * class_counts[i] for i in range(num_classes)) / total_samples


Validation Accuracy: 0.7122471098265896
Class 0: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 1: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 2: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 3: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 4: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 5: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 6: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 7: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 8: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 9: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 10: Precision = 0.8191, Recall = 0.8463, F1-score = 0.8325
Class 11: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 12: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 13: Precision = 0.0000, Recall = 0.0000, F1-score = 0.0000
Class 14: Precision = 0.0000, Recall = 0.0000, F1-score = 0.

In [35]:
print(f"Weighted Average F1-score: {weighted_f1:.4f}")

Weighted Average F1-score: 0.6997
